pkg/bucket/bandwidth/throttler.go

```diff
@@ -0,0 +1,115 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bandwidth
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Throttler implements the throttling for bandwidth
+type Throttler struct {
+	generateTicker         *time.Ticker // Ticker to generate available bandwidth
+	freeBytes              int64        // unused bytes in the interval
+	bytesPerInterval       int64        // bytes allocated for the interval
+	intervalInMilliSeconds int64        // interval to check bandwidth consumption and throttle threads in
+	lock                   sync.Mutex   // lock
+}
+
+// NewThrottler returns a new bandwidth throttler. Set bytesPerSecond to 0 for no limit
+func NewThrottler(bytesPerSecond int64) *Throttler {
+	interval := 100 // interval for which to block requests
+	t := &Throttler{
+		generateTicker:         time.NewTicker(time.Duration(interval) * time.Millisecond),
+		intervalInMilliSeconds: int64(interval),
+	}
+	t.SetBandwidth(bytesPerSecond)
+	t.freeBytes = t.bytesPerInterval
+	go t.generateBandwidth()
+	return t
+}
+
+// GetLimitForBytes gets the bytes that are possible to send within the limit
+// if want is <= 0 or no bandwidth limit set, returns want
+func (t *Throttler) GetLimitForBytes(want int64) int64 {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	var send int64
+	if want <= 0 || atomic.LoadInt64(&t.bytesPerInterval) == 0 {
+		return want
+	}
+	for {
+		if atomic.LoadInt64(&t.freeBytes) >= want {
+			send = want
+		} else {
+			send = atomic.LoadInt64(&t.freeBytes)
+		}
+		if send == 0 {
+			time.Sleep(time.Duration(t.intervalInMilliSeconds/1) * time.Millisecond)
+			continue
+		}
+		atomic.AddInt64(&t.freeBytes, -send)
+		return send
+	}
+}
+
+// WaitToSend blocks till there is spare bandwidth
+func (t *Throttler) WaitToSend(bytes int64) {
+	// 0 == infinite bandwidth
+	if atomic.LoadInt64(&t.bytesPerInterval) == 0 || bytes == 0 {
+		return
+	}
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	for {
+		delta := atomic.LoadInt64(&t.freeBytes) - bytes
+		if delta < 0 {
+			time.Sleep(time.Duration(t.intervalInMilliSeconds/5) * time.Millisecond)
```

Comment:
```
// WaitToSend blocks till there is spare bandwidth
func (t *Throttler) WaitToSend(bytes int64) {
	// 0 == infinite bandwidth
	if bytes == 0 || atomic.LoadInt64(&t.bytesPerInterval) == 0 {
		return
	}
	t.lock.Lock()
	defer t.lock.Unlock()
	for {
		delta := atomic.LoadInt64(&t.freeBytes) - bytes
		if delta < 0 {
			t.signal.Wait()
			continue
		}
		atomic.AddInt64(&t.freeBytes, -bytes)
		return
	}
}
```

Add `signal *sync.Cond` to `Throttle`. Add `t.signal.Broadcast()` to wake up one of the waiters after adding bandwidth. This could also grab a lock for even more security, otherwise requestors may miss a cycle.